hwdef: ARKV6S: add board support (#33157)
ArduPilot/ardupilot: libraries/AP_HAL_ChibiOS/hwdef/ARKV6S/README.md

# ARKV6S Flight Controller

[ARK Electronics](https://arkelectron.com/product/arkv6s/)

The ARKV6S is a low-cost, single-IMU variant of the [ARKV6X](../ARKV6X), based on the [Pixhawk Autopilot Bus (PAB) Form Factor](https://github.com/pixhawk/Pixhawk-Standards/blob/master/DS-010%20Pixhawk%20Autopilot%20Bus%20Standard.pdf). It plugs into any PAB-compatible carrier board.

## Features

### Processor

- STM32H743IIK6 32-bit processor
- 480MHz
- 2MB Flash
- 1MB RAM

### Sensors

- Invensense IIM-42653 Industrial IMU with heater resistor
- Bosch BMP390 Barometer
- ST IIS2MDC Magnetometer
- FRAM (for parameter storage)

### Power

- 5V input
- 500mA (300mA for main system, 200mA for heater)

### Form Factor

- [Pixhawk Autopilot Bus (PAB) Form Factor](https://github.com/pixhawk/Pixhawk-Standards/blob/master/DS-010%20Pixhawk%20Autopilot%20Bus%20Standard.pdf)
- Size: 36 x 29 x 5 mm
- Weight: 5.0 g

## Pinout

> The ARKV6S pinout conforms to the [Pixhawk V6X standard](https://github.com/pixhawk/Pixhawk-Standards/blob/master/DS-010%20Pixhawk%20Autopilot%20Bus%20Standard.pdf).

Reviewer: needs exact pinout, conforming can mean pins listed are not pinned out. Also you dont specify which variant or its version....a user would not have a clue....will not review further without an image or pinout table

# ARKV6S Flight Controller

[ARK Electronics](https://arkelectron.com/product/arkv6s/)

The ARKV6S is a low-cost, single-IMU variant of the [ARKV6X](../ARKV6X), based on the [Pixhawk Autopilot Bus (PAB) Form Factor](https://github.com/pixhawk/Pixhawk-Standards/blob/master/DS-010%20Pixhawk%20Autopilot%20Bus%20Standard.pdf). It plugs into any PAB-compatible carrier board.

## Features

### Processor

- STM32H743IIK6 32-bit processor
- 480MHz
- 2MB Flash
- 1MB RAM

### Sensors

- Invensense IIM-42653 Industrial IMU with heater resistor
- Bosch BMP390 Barometer
- ST IIS2MDC Magnetometer
- FRAM (for parameter storage)

### Power

- 5V input
- 500mA (300mA for main system, 200mA for heater)

### Form Factor

- [Pixhawk Autopilot Bus (PAB) Form Factor](https://github.com/pixhawk/Pixhawk-Standards/blob/master/DS-010%20Pixhawk%20Autopilot%20Bus%20Standard.pdf)
- Size: 36 x 29 x 5 mm
- Weight: 5.0 g

![ARKV6S](arkv6s.png)

## Pinout

The ARKV6S pinout conforms to the [Pixhawk V6X standard](https://github.com/pixhawk/Pixhawk-Standards/blob/master/DS-010%20Pixhawk%20Autopilot%20Bus%20Standard.pdf).

![PinOut](pinout.png)

## UART Mapping

> | Name    | Function              |

Reviewer: ```suggestion
| Name    | UART/Protocol              |
```

|:--------|:----------------------|
| SERIAL0 | USB                   |

> | SERIAL1 | UART7 (Telem1)        |

Reviewer: ```suggestion
| SERIAL1 | UART7 (Telem1/MAVLink2)        |
```

> | SERIAL2 | UART5 (Telem2)        |

Reviewer: ```suggestion
| SERIAL2 | UART5 (Telem2/MAVLink2)        |
```

| SERIAL3 | USART1 (GPS1)         |
| SERIAL4 | UART8 (GPS2)          |

> | SERIAL5 | USART2 (Telem3)       |

Reviewer: ```suggestion
| SERIAL5 | USART2 (Telem3/None)       |
```

> | SERIAL6 | UART4                 |

Reviewer: ```suggestion
| SERIAL6 | UART4/None                 |
```

> | SERIAL7 | USART3 (Debug)        |

Reviewer: ```suggestion
| SERIAL7 | USART3 (Debug/None)        |
```

> | SERIAL8 | USART6 (RC Input)     |

Reviewer: ```suggestion
| SERIAL8 | USART6 (RCIN)     |
```

| SERIAL9 | OTG2 (MAVLink2)       |

All UARTs support DMA. Any UART may be re-tasked by changing its protocol parameter. The Telem1, Telem2, and Telem3 ports have RTS/CTS pins, the other UARTs do not have RTS/CTS.

> USART6 may alternatively be wired to an IOMCU on a carrier board. To use it for the IOMCU instead of as a direct serial RC input, edit `hwdef.dat` to uncomment the `IOMCU_UART USART6` line and switch to the alternate `SERIAL_ORDER` line.

Reviewer: ```suggestion
USART6 may alternatively be wired to an IOMCU on a carrier board. To use it for the IOMCU instead of as a direct serial RC input, edit `hwdef.dat` to uncomment the `IOMCU_UART USART6` line and switch to the alternate `SERIAL_ORDER` line. You must be able to build firmware locally to do this.
```

## CAN

The ARKV6S exposes two independent CAN buses on the PAB connector. The PHY is held in reset until firmware enables it. Both CAN drivers are enabled by default (`CAN_P1_DRIVER` = `1`, `CAN_P2_DRIVER` = `1`) for DroneCAN peripherals.

## Ethernet

The ARKV6S has built-in 100 Mbit Ethernet (RMII PHY: Microchip LAN8742A). The PHY is held in reset until firmware enables it via GPIO 113. The RMII signals are routed to the PAB connector for the carrier board to break out to a connector.

## microSD

A microSD socket on the carrier board is supported with the FATFS filesystem, used for parameter and dataflash logging.

## RC Input

> By default, RC input is configured on USART6 (`SERIAL8_PROTOCOL` = `23`). It supports all RC protocols except PPM. A separate timer input (TIM8_CH1) is also available for unidirectional RC protocols. See [Radio Control Systems](https://ardupilot.org/copter/docs/common-rc-systems.html) for details for a specific RC system.

Reviewer: ```suggestion
By default, RC input is configured on USART6  RX (labeled UASART6_RX_). It supports all unidirectional RC protocols except PPM.  Bi-Directional protocols (CRSF/ELRS,etc.) require connection the the UART6 TX (labeled USART6_TX_TO_IO_NC).See [Radio Control Systems](https://ardupilot.org/copter/docs/common-rc-systems.html) for details for a specific RC system

A separate timer input (FMU_PPM_INPUT_FROM_IO_RC_IN) is also available for unidirectional RC protocols, including PPM. .
```

## Battery Monitoring

The board supports an INA226 digital power monitor on I2C1 (`HAL_BATT_MONITOR_DEFAULT = 21`) by default as the first monitor. Additional analog voltage/current inputs may be available on the carrier board used for this module.

## Compass

This autopilot has a built-in compass. The compass is the IIS2MDC. Often this internal compass is disabled due to power interference and a remotely located compass is used.

## IMU Heater

The ARKV6S has a 1W heater that keeps the IIM-42653 IMU warm in extreme conditions. The heater target temperature defaults to 45 °C and can be tuned with `BRD_IMUHEAT_*` parameters.

## PWM Output

> PWM outputs are provided through the module connector. PWM 1-6 are capable of PWM and DShot (including bi-directional DShot).  PWM 7-8 are PWM-only since they have no DMA. Outputs in the same timer group must use the same protocol:

Reviewer: ```suggestion
PWM outputs are provided through the module connector(labeled FMU_CHx). PWM (CHx) 1-6 are capable of PWM and DShot (including bi-directional DShot).  PWM (CHx) 7-8 are PWM-only since they have no DMA. Outputs in the same timer group must use the same protocol:
```

- PWM 1-4 Group1 (TIM5)
- PWM 5-6 Group2 (TIM4)
- PWM 7-8 Group3 (TIM12, no DMA)

## Loading Firmware

> 

Reviewer: certainly it comes with either PX4 or AP bootloader installed....which one?probably PX4....which should still load AP firmware via GCS....we should state this, give info on where the firmware is located, ie folder name and server url...
THEN state that the AP bootloader may be flashed using ST-LINK.
Also, that they change flash the AP bootloader AFTER AP firmware is loaded by having the GCS update the bootloader.

Author: I've updated the language, let me know what you think. I've also removed the
```
define AP_BOOTLOADER_FLASHING_ENABLED 0
```
so that the bootloader gets embedded in the image and live BL update is available

The bootloader is flashed with an ST-Link. Pre-built bootloader binaries can be found at the [ArduPilot firmware server](https://firmware.ardupilot.org/Tools/Bootloaders/).

```bash
st-flash write ARKV6S_bl.bin 0x08000000
```

> Once the bootloader is installed you can update the firmware using any ArduPilot ground station software. Firmware can be found at the [ArduPilot firmware server](https://firmware.ardupilot.org/) in folders labeled "ARKV6S". Updates should be done with the `*.apj` firmware files.

Reviewer: you can load AP firmware with PX4 bootloader...this is not needed  unless you are shipping this module with no bootloader (unlikely I would think)

Author: I've updated this section

# ARKV6S Flight Controller

[ARK Electronics](https://arkelectron.com/product/arkv6s/)

The ARKV6S is a low-cost, single-IMU variant of the [ARKV6X](../ARKV6X), based on the [Pixhawk Autopilot Bus (PAB) Form Factor](https://github.com/pixhawk/Pixhawk-Standards/blob/master/DS-010%20Pixhawk%20Autopilot%20Bus%20Standard.pdf). It plugs into any PAB-compatible carrier board.

## Features

### Processor

- STM32H743IIK6 32-bit processor
- 480MHz
- 2MB Flash
- 1MB RAM

### Sensors

- Invensense IIM-42653 Industrial IMU with heater resistor
- Bosch BMP390 Barometer
- ST IIS2MDC Magnetometer
- FRAM (for parameter storage)

### Power

- 5V input
- 500mA (300mA for main system, 200mA for heater)

### Form Factor

- [Pixhawk Autopilot Bus (PAB) Form Factor](https://github.com/pixhawk/Pixhawk-Standards/blob/master/DS-010%20Pixhawk%20Autopilot%20Bus%20Standard.pdf)
- Size: 36 x 29 x 5 mm
- Weight: 5.0 g

![ARKV6S](bottom.png)

> ## Pinout

Reviewer: you did not address the image issues I mentioned  in previous review

Author: > is PH11 (which is a PWM output) really connected to pin 50 of X2?

Yes

> HWVER pins are not defined...should be removed since AP does not use them anyway

This signal is present on the connector, the circuit exists on the board. The png reflects the reality of the interface.

Reviewer: but will be confusing to a normal AP user, should just state that the VER pins on X1 have no function, and that PH11 on X2 is actually PWM3

Author: Oh hold on, this is a typo in the schematic that got carried into the png. The net has two labels, the correct label is PD15. Will fix. I will also add a comment about VER being unused.

Author: Corrected the png abd added a note below it
<img width="483" height="51" alt="image" src="https://github.com/user-attachments/assets/3ffdd873-6c20-4c5f-9e79-2bf0b238754f" />
> _**Note:** HW_VER_SENSE and HW_VER_REV_DRIVE are unused by Ardupilot._

Author: The confusion comes from poorly labelled nets in the official PAB spec. Most have generic labels (eg FMU_CH1, BUZZER_1, etc) which doesn't enforce a function<-->IO mapping. In the spec pin 50 is named "PH11", however it's actually connected to PD15 on the H7. It seems the naming wasn't fully thought out in the spec.
[DS-010 Pixhawk Autopilot Bus Standard.pdf](https://github.com/user-attachments/files/29857047/DS-010.Pixhawk.Autopilot.Bus.Standard.pdf)

Reviewer: Understood...then PD15 is undefined and should be added to the note

# ARKV6S Flight Controller

[ARK Electronics](https://arkelectron.com/product/arkv6s/)

The ARKV6S is a low-cost, single-IMU variant of the [ARKV6X](../ARKV6X), based on the [Pixhawk Autopilot Bus (PAB) Form Factor](https://github.com/pixhawk/Pixhawk-Standards/blob/master/DS-010%20Pixhawk%20Autopilot%20Bus%20Standard.pdf). It plugs into any PAB-compatible carrier board.

## Features

### Processor

- STM32H743IIK6 32-bit processor
- 480MHz
- 2MB Flash
- 1MB RAM

### Sensors

- Invensense IIM-42653 Industrial IMU with heater resistor
- Bosch BMP390 Barometer
- ST IIS2MDC Magnetometer
- FRAM (for parameter storage)

### Power

- 5V input
- 500mA (300mA for main system, 200mA for heater)

### Form Factor

- [Pixhawk Autopilot Bus (PAB) Form Factor](https://github.com/pixhawk/Pixhawk-Standards/blob/master/DS-010%20Pixhawk%20Autopilot%20Bus%20Standard.pdf)
- Size: 36 x 29 x 5 mm
- Weight: 5.0 g

## Pinout

> The ARKV6S is a PAB module. For the connector and signal pinout, see the [DS-10 Pixhawk Autopilot Bus Standard](https://github.com/pixhawk/Pixhawk-Standards/blob/master/DS-010%20Pixhawk%20Autopilot%20Bus%20Standard.pdf).

Reviewer: which pinout does the module conform to? 6X, 6C,etc.? better to put the exact pinout in readme...see https://github.com/ArduPilot/ardupilot/tree/master/libraries/AP_HAL_ChibiOS/hwdef/CUAV-V6X-v2 for an example, including formats of UART and PWM sections

## UART Mapping

| Name    | Function              |
|:--------|:----------------------|
| SERIAL0 | USB                   |
| SERIAL1 | UART7 (Telem1)        |
| SERIAL2 | UART5 (Telem2)        |
| SERIAL3 | USART1 (GPS1)         |
| SERIAL4 | UART8 (GPS2)          |
| SERIAL5 | USART2 (Telem3)       |
| SERIAL6 | UART4                 |
| SERIAL7 | USART3 (Debug)        |
| SERIAL8 | USART6 (RC Input)     |
| SERIAL9 | OTG2 (MAVLink2)       |

> All UARTs support DMA. Any UART may be re-tasked by changing its protocol parameter.

Reviewer: no mention of flow control capabilities on UART2/5/7

USART6 may alternatively be wired to an IOMCU on a carrier board. To use it for the IOMCU instead of as a direct serial RC input, edit `hwdef.dat` to uncomment the `IOMCU_UART USART6` line and switch to the alternate `SERIAL_ORDER` line.

## CAN

The ARKV6S exposes two independent CAN buses on the PAB connector:

| Bus  | TX  | RX  |
|:-----|:----|:----|
| CAN1 | PD1 | PD0 |
| CAN2 | PB13| PB12|

Both CAN drivers are enabled by default (`CAN_P1_DRIVER` = `1`, `CAN_P2_DRIVER` = `1`) for DroneCAN peripherals.

## Ethernet

The ARKV6S has built-in 100 Mbit Ethernet (RMII PHY: Microchip LAN8742A). The PHY is held in reset until firmware enables it via GPIO 113. The RMII signals are routed to the PAB connector for the carrier board to break out to a connector.

## microSD

A microSD socket on the carrier board is supported via SDMMC2 with the FATFS filesystem, used for parameter and dataflash logging.

## RC Input

> By default, RC input is configured on USART6 (`SERIAL8_PROTOCOL` = `23`). It supports all RC protocols except PPM. See [Radio Control Systems](https://ardupilot.org/copter/docs/common-rc-systems.html) for details for a specific RC system.

Reviewer: While UART6 is setup for RC, you also have an RCINT input defined for unidirectional protocols, which is not detailed in this section.

## Battery Monitoring

> The board supports an INA226 digital power monitor on I2C1 (`HAL_BATT_MONITOR_DEFAULT = 21`). Additional analog voltage/current inputs are available on the carrier board.

Reviewer: ```suggestion
The board supports an INA226 digital power monitor on I2C1 (`HAL_BATT_MONITOR_DEFAULT = 21`) by default as the first monitor. Additional analog voltage/current inputs may be available on the carrier board used for this module.
```

## Compass

This autopilot has a built-in compass. The compass is the IIS2MDC. Often this internal compass is disabled due to power interference and a remotely located compass is used.

## IMU Heater

The ARKV6S has a 1W heater that keeps the IIM-42653 IMU warm in extreme conditions. The heater target temperature defaults to 45 °C and can be tuned with `BRD_IMUHEAT_*` parameters.

## PWM Output

> PWM outputs are provided through the PAB connector. PWM 1-6 are capable of PWM and DShot (including bi-directional DShot); PWM 7-8 are PWM-only since TIM12 has no DMA. Outputs in the same timer group must use the same protocol:

Reviewer: ```suggestion
PWM outputs are provided through the module connector. PWM 1-6 are capable of PWM and DShot (including bi-directional DShot).  PWM 7-8 are PWM-only since they have no DMA. Outputs in the same timer group must use the same protocol:
```

- PWM 1-4 Group1 (TIM5)
- PWM 5-6 Group2 (TIM4)
- PWM 7-8 Group3 (TIM12, no DMA)

## Loading Firmware

> Initial firmware load can be done with DFU by plugging in USB with the BOOT button pressed. You can then load the bootloader using your favorite DFU tool. The bootloader can be found at the [ArduPilot firmware server](https://firmware.ardupilot.org/Tools/Bootloaders/).

Reviewer: ```suggestion
Initial firmware load can be done with DFU by plugging in USB with the BOOT button pressed. You can then load the firmware using your favorite DFU tool. The firmware can be found at the [ArduPilot firmware server](https://firmware.ardupilot.org/) in folders labeled "ARKV6S".
```

Once the initial firmware is loaded you can update the firmware using any ArduPilot ground station software. Updates should be done with the `*.apj` firmware files.

> Alternatively you can build the firmware from source:

Reviewer: thiis section is not needed
